### ISSUE TITLE
Some issues when reproducing the results

### ISSUE THREAD

**Issue by Chaolei98** [benedettaliberatori/T3AL#4]
A great job! Thanks to the author for opening the source code. However I have some problems when reproducing the results. Any help from the authors would be appreciated:
(1) When I run the code on thumos dataset with 50 split setting, the results of training free experiment have some gap with the paper listed.
(2) tt_method lacks captions file when using text-guided region suppression.
(3) Why the value passed into output["label"] in tt_method.py / tf_method.py is the pseudo label index rather than the updated region label as mentioned in the paper？
![Q1](https://github.com/benedettaliberatori/T3AL/assets/67382837/168e6b73-69b1-4509-8583-df344dd8cef0)

**Comment by benedettaliberatori** [comment]
Hi, thank you for your interest in our work and for your questions!

**1.** The problem you are encountering with the training-free baseline is probably due to the features you are using. The features provided via Google Drive are extracted with [CoCa](https://github.com/mlfoundations/open_clip/blob/main/src/open_clip/coca_model.py) up to the last projection layer. This is because we do not propagate gradients before that in our method. For the training-free baseline you need to use the features obtained after the projection. I was using already extracted ones, now I have added the projection in the [forward](https://github.com/benedettaliberatori/T3AL/blob/1bc3a0cf269072ea45f09fb6e8ac64076de4ca43/src/models/components/tf_method.py#L205) so you can use the same features for baseline and method. 

```
with torch.no_grad():
     image_features = image_features @ self.model.visual.proj 
```

**2.** I have added the directories with the generated captions for both datasets on Google Drive. You can find the links in the updated README. You can download the files and change the paths accordingly. 

**3.** Yes, but results-wise is the same. If you substitute `indexes.item()` with `pred[i]`, you can check it. Once the pseudo-label is computed, after the TTA steps, the intervals are selected based on the cosine similarity with the pseudo-label embedding, so the class of the pseudo-label has a higher probability. 


Hope this clears your doubts!

**Comment by Chaolei98** [comment]
Thank you for your reply! The Q2 and Q3 have been solved. Following the projection solution in Q1,  performance has improved but still far from the paper. Specifically, on the THUMOS14 dataset, under the 50/50 split strategy, the average AP is 0.18 for the training free baseline (tf_method) and 0.616 for the main method (tt_method), whereas the corresponding results for the paper are 4.8 and 10.4 (as listed in Table 1 of the paper).  Although I only show the results with split=0 instead of average value after 10 different splits, I don't think  this is the reason for such a large performance gap, which is something I'd really like to know. I only changed the code for getting fps from video files by annotations[video_name]['duration_frame'] / annotations[video_name]['duration_second'] due to lack of video path ( I used pre-extracted CoCa features to speed up the inference instead of the raw video). Is there a problem with my implementation? I'd be grateful if you could enlighten me.

**Comment by benedettaliberatori** [comment]
Yes, I guess that is what is causing the mismatch. 
Do not rely on the `duration_frame` for the fps computation. We decode videos at the original fps, this is why we use the `get_video_fps` function to get the number from videos. If you compute fps from the json file you get 10. If you do not want to download videos to get these numbers, consider that most of the videos in THUMOS14 have 30 fps, so you can use this number as a proxy for the whole dataset. Yet, I advise you to download the dataset in video format as well if you can. 

Let me know if this solves your problem!

**Comment by Chaolei98** [comment]
Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.

**Comment by hanjiang2001** [comment]
> Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.

Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.

**Comment by Chaolei98** [comment]
> > Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.
> 
> Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.

I downloaded the original dataset. Using 30 fps can yield approximately similar results. Another option is to look for other open-source repositories that use the THUMOS14 or ActivityNet datasets—some of them might provide JSON files containing the exact fps values！

**Comment by hanjiang2001** [comment]
> > > Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.
> > 
> > 
> > Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.
> 
> I downloaded the original dataset. Using 30 fps can yield approximately similar results. Another option is to look for other open-source repositories that use the THUMOS14 or ActivityNet datasets—some of them might provide JSON files containing the exact fps values！

Thank you very much for your reply! It has helped me a lot.

**Comment by hanjiang2001** [comment]
> > > Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.
> > 
> > 
> > Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.
> 
> I downloaded the original dataset. Using 30 fps can yield approximately similar results. Another option is to look for other open-source repositories that use the THUMOS14 or ActivityNet datasets—some of them might provide JSON files containing the exact fps values！

I'm really sorry to bother you again. After obtaining the FPS using the original videos, I encountered some issues when trying to reproduce the results. On the **Results on THUMOS14 (50%-50%)**, my result was **10.14**, which is close to the one reported in the paper. However, on the **Results on THUMOS14 (75%-25%)**, I could only achieve a score of **5.86**, which is far from the result stated in the paper.
I believe there might be an error in my script, because under the 50-50 setting, I obtained reasonable results. Were you able to successfully reproduce these results at that time? Could you please provide me with valuable assistance? Thank you very much!
The scripts I used are as follows:


#!/bin/bash
EXPERIMENT="tt_thumos"
DATA="thumos"
SETTING="75"
VPATH='videos/THUMOS14/all'

for i in $(seq 0 9)
do
  echo "model.split=$i data.nsplit=$i"
  python src/train.py \
    experiment=$EXPERIMENT \
    data=$DATA \
    model.setting=$SETTING \
    model.split=$i \
    data.nsplit=$i \
    model.video_path=$VPATH
done

**Comment by benedettaliberatori** [comment]
Thanks @Chaolei98 for the fast reply! @EricPaul03, I replied to you in #20.

**Comment by Chaolei98** [comment]
> > > > Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.
> > > 
> > > 
> > > Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.
> > 
> > 
> > I downloaded the original dataset. Using 30 fps can yield approximately similar results. Another option is to look for other open-source repositories that use the THUMOS14 or ActivityNet datasets—some of them might provide JSON files containing the exact fps values！
> 
> I'm really sorry to bother you again. After obtaining the FPS using the original videos, I encountered some issues when trying to reproduce the results. On the **Results on THUMOS14 (50%-50%)**, my result was **10.14**, which is close to the one reported in the paper. However, on the **Results on THUMOS14 (75%-25%)**, I could only achieve a score of **5.86**, which is far from the result stated in the paper. I believe there might be an error in my script, because under the 50-50 setting, I obtained reasonable results. Were you able to successfully reproduce these results at that time? Could you please provide me with valuable assistance? Thank you very much! The scripts I used are as follows:
> 
> #!/bin/bash EXPERIMENT="tt_thumos" DATA="thumos" SETTING="75" VPATH='videos/THUMOS14/all'
> 
> for i in $(seq 0 9) do echo "model.split=$i data.nsplit=$i" python src/train.py experiment=$EXPERIMENT data=$DATA model.setting=$SETTING model.split=$i data.nsplit=$i model.video_path=$VPATH done

I successfully reproduced the results in the paper both in the 75-25 and 50-50 settings.
I didn’t encounter the issue you mentioned, but one thing worth noting is that the hyperparameter `split` in `/T3AL/config/thumos.yaml` also needs to be manually set to 75 instead of 50.
Could this be the reason for the poor performance? I’m quite curious as well, since I checked your command and it looks correct to me.
Please let me know if you manage to solve the issue!

**Comment by hanjiang2001** [comment]
> > > > > Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.
> > > > 
> > > > 
> > > > Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.
> > > 
> > > 
> > > I downloaded the original dataset. Using 30 fps can yield approximately similar results. Another option is to look for other open-source repositories that use the THUMOS14 or ActivityNet datasets—some of them might provide JSON files containing the exact fps values！
> > 
> > 
> > I'm really sorry to bother you again. After obtaining the FPS using the original videos, I encountered some issues when trying to reproduce the results. On the **Results on THUMOS14 (50%-50%)**, my result was **10.14**, which is close to the one reported in the paper. However, on the **Results on THUMOS14 (75%-25%)**, I could only achieve a score of **5.86**, which is far from the result stated in the paper. I believe there might be an error in my script, because under the 50-50 setting, I obtained reasonable results. Were you able to successfully reproduce these results at that time? Could you please provide me with valuable assistance? Thank you very much! The scripts I used are as follows:
> > #!/bin/bash EXPERIMENT="tt_thumos" DATA="thumos" SETTING="75" VPATH='videos/THUMOS14/all'
> > for i in $(seq 0 9) do echo "model.split=$i data.nsplit=$i" python src/train.py experiment=$EXPERIMENT data=$DATA model.setting=$SETTING model.split=$i data.nsplit=$i model.video_path=$VPATH done
> 
> I successfully reproduced the results in the paper both in the 75-25 and 50-50 settings. I didn’t encounter the issue you mentioned, but one thing worth noting is that the hyperparameter `split` in `/T3AL/config/thumos.yaml` also needs to be manually set to 75 instead of 50. Could this be the reason for the poor performance? I’m quite curious as well, since I checked your command and it looks correct to me. Please let me know if you manage to solve the issue!

Thank you very much! I think you may have indeed solved my problem! This was an oversight on my part, and I will try this modification as soon as possible.

**Comment by Chaolei98** [comment]
> > > > > > Thanks for the suggestion. Results by using fps 30 is much better than using `duration_frame` before. I think the exact fps alternative approximation for each video can further improve performance. You solved my confusion, thanks again.
> > > > > 
> > > > > 
> > > > > Hello, I've also encountered a similar issue. Downloading the original videos can be quite cumbersome. I'm wondering how you eventually resolved this. Did you simply use a fixed frame rate of 30fps for both datasets? Does this approach yield results comparable to those reported in the paper? I would really appreciate it if you could share your settings and insights.
> > > > 
> > > > 
> > > > I downloaded the original dataset. Using 30 fps can yield approximately similar results. Another option is to look for other open-source repositories that use the THUMOS14 or ActivityNet datasets—some of them might provide JSON files containing the exact fps values！
> > > 
> > > 
> > > I'm really sorry to bother you again. After obtaining the FPS using the original videos, I encountered some issues when trying to reproduce the results. On the **Results on THUMOS14 (50%-50%)**, my result was **10.14**, which is close to the one reported in the paper. However, on the **Results on THUMOS14 (75%-25%)**, I could only achieve a score of **5.86**, which is far from the result stated in the paper. I believe there might be an error in my script, because under the 50-50 setting, I obtained reasonable results. Were you able to successfully reproduce these results at that time? Could you please provide me with valuable assistance? Thank you very much! The scripts I used are as follows:
> > > #!/bin/bash EXPERIMENT="tt_thumos" DATA="thumos" SETTING="75" VPATH='videos/THUMOS14/all'
> > > for i in $(seq 0 9) do echo "model.split=$i data.nsplit=$i" python src/train.py experiment=$EXPERIMENT data=$DATA model.setting=$SETTING model.split=$i data.nsplit=$i model.video_path=$VPATH done
> > 
> > 
> > I successfully reproduced the results in the paper both in the 75-25 and 50-50 settings. I didn’t encounter the issue you mentioned, but one thing worth noting is that the hyperparameter `split` in `/T3AL/config/thumos.yaml` also needs to be manually set to 75 instead of 50. Could this be the reason for the poor performance? I’m quite curious as well, since I checked your command and it looks correct to me. Please let me know if you manage to solve the issue!
> 
> Thank you very much! I think you may have indeed solved my problem! This was an oversight on my part, and I will try this modification as soon as possible.

No problem at all! I’m glad the suggestion is helpful.
